File: backend/app/core/password_rules.py
"""Règles d’inscription / reset. Le login des anciens hash n’est pas concerné."""

import re

MIN_LENGTH = 8
MAX_LENGTH = 72

PASSWORD_RULE_MESSAGE = (
    "Le mot de passe doit contenir au moins 8 caractères, une minuscule, "
    "une MAJUSCULE, un chiffre et un caractère spécial."
)

_LOWER = re.compile(r"[a-z]")
_UPPER = re.compile(r"[A-Z]")
_DIGIT = re.compile(r"\d")
_SPECIAL = re.compile(r"[^A-Za-z0-9]")


def password_strength_error(password: str) -> str | None:
    if len(password) < MIN_LENGTH or len(password) > MAX_LENGTH:
        return PASSWORD_RULE_MESSAGE
    if _LOWER.search(password) is None:
        return PASSWORD_RULE_MESSAGE
    if _UPPER.search(password) is None:
        return PASSWORD_RULE_MESSAGE
    if _DIGIT.search(password) is None:
        return PASSWORD_RULE_MESSAGE
    if _SPECIAL.search(password) is None:
        return PASSWORD_RULE_MESSAGE
    return None


def require_strong_password(password: str) -> str:
    error = password_strength_error(password)
    if error is not None:
        raise ValueError(error)
    return password

File: lib/core/auth/password_rules.dart
/// Règles d’inscription / reset. Identiques au backend.
///
/// Le login des comptes déjà hashés (ex. démo `motdepasse`) n’est pas
/// soumis à cette règle.
class PasswordRules {
  static const minLength = 8;
  static const maxLength = 72;

  static const message =
      'Le mot de passe doit contenir au moins 8 caractères, une minuscule, '
      'une MAJUSCULE, un chiffre et un caractère spécial.';

  static String? validate(String? value) {
    final password = value ?? '';
    if (password.length < minLength || password.length > maxLength) {
      return message;
    }
    if (!RegExp(r'[a-z]').hasMatch(password)) {
      return message;
    }
    if (!RegExp(r'[A-Z]').hasMatch(password)) {
      return message;
    }
    if (!RegExp(r'\d').hasMatch(password)) {
      return message;
    }
    if (!RegExp(r'[^A-Za-z0-9]').hasMatch(password)) {
      return message;
    }
    return null;
  }

  static bool isStrong(String value) => validate(value) == null;
}
